ui-components/src/page_components/listing/ListingCard.tsx

```diff
@@ -22,10 +25,75 @@ export interface ListingCardProps {
   tableHeaderProps?: ListingCardHeaderProps
   tableProps: ListingCardTableProps
   detailsLinkClass?: string
+  listing: Listing
 }
 
 const ListingCard = (props: ListingCardProps) => {
   const { imageCardProps, tableProps, detailsLinkClass, tableHeaderProps } = props
+  const { profile, userProfileService } = useContext(AuthContext)
+  const [updatedFavorites, setUpdatedFavorites] = useState(profile?.preferences?.favoriteIDs)
+
+  useEffect(() => {
+    setFavoriteButtonState()
+  })
+
+  const setFavoriteButtonState = () => {
+    const preferences: UserPreferences = profile?.preferences || { favoriteIDs: [] }
+    setUpdatedFavorites(preferences?.favoriteIDs)
+  }
+
+  const addToFavorite = async () => {
+    if (!profile) {
+      return
+    }
+    const preferences: UserPreferences = profile?.preferences || { favoriteIDs: [] }
+    if (!preferences.favoriteIDs) {
+      preferences.favoriteIDs = []
+    }
+
+    if (!preferences.favoriteIDs.includes(props.listing.id)) {
+      preferences.favoriteIDs.push(props.listing.id)
+      console.log("Adding Listing")
+    } else {
+      return
+    }
+
+    try {
+      await userProfileService?.update({
+        body: { ...profile, preferences },
+      })
+      setUpdatedFavorites(updatedFavorites?.concat(props.listing.id))
+    } catch (err) {
+      console.warn(err)
+    }
+  }
+
+  const removeFavorite = async () => {
+    if (!profile) {
+      return
+    }
+    const preferences: UserPreferences = profile?.preferences || { favoriteIDs: [] }
+    if (!preferences.favoriteIDs || preferences.favoriteIDs.length === 0) {
+      return
+    }
+
+    const index: number = preferences.favoriteIDs.indexOf(props.listing.id)
+    const temp = [
```

Comment:
Using `splice` is a little more concise:

```
preferences.favoriteIDs.splice(index, 1)
```

It also modifies it in place, so you can remove the `temp` variable.

Reply:
modified